src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs

```diff
@@ -114,30 +119,49 @@ public BodyModelBinder(IList<IInputFormatter> formatters, IHttpRequestStreamRead
             if (formatter == null)
             {
                 _logger?.NoInputFormatterSelected(formatterContext);
+
                 var message = Resources.FormatUnsupportedContentType(httpContext.Request.ContentType);
                 var exception = new UnsupportedContentTypeException(message);
-                bindingContext.ModelState.AddModelError(modelBindingKey, exception, bindingContext.ModelMetadata);
+                if (bindingContext.IsTopLevelObject &&
+                    !string.Equals(bindingContext.ModelName, modelBindingKey, StringComparison.Ordinal))
+                {
+                    // Cannot use ModelStateEntry tracked in ModelBindingContext due to "special logic" above.
+                    modelState.TryAddModelError(modelBindingKey, exception, metadata);
+                }
+                else
+                {
+                    bindingContext.ModelStateEntry.TryAddModelError(modelState, modelBindingKey, exception, metadata);
+                }
```

Comment:
This is about the ugliest change need to use the new API. Happens once more in this class (or as @Eilon overwhelmingly prefers `this` `class`) and again in `FormFileModelBinder` -- the two most special-casey binders.